libcalico-go/lib/backend/k8s/conversion/conversion.go

```diff
@@ -809,3 +823,15 @@ func stringsToIPNets(ipStrings []string) ([]*cnet.IPNet, error) {
 	}
 	return podIPNets, nil
 }
+
+// ConvertUID converts a UID to a new UID in a deterministic way. This is useful when we want to generate a new UID
+// for a resource that is derived from another resource, but we don't want to use the same UID in order to
+// ensure that the new resource is treated as unique. This is important, as two objects with the same UID causes
+// confusion in the Kubernetes garbage collection logic.
+func ConvertUID(uid types.UID) (types.UID, error) {
+	newUID, err := uuid.NewRandomFromReader(strings.NewReader(string(uid)))
```

Comment:
My big concern with this is that NewRandomFromReader loses randomness, as it's simply reading the input bytes into a new UUID and forcing the Version+Variant bits to the correct values. IIUC, it's really not too different from Rob's original approach of reordering the using the bits from the original UUID except that it loses more randomness because the non-random `-` characters are included in the resulting UUID.

We could also go with the original approach that Rob used by reversing the values and then asserting the Version+Variant bits in the same way that NewRandomFromReader does.

Either way, this function should only end up generating UIDs for:
- Profiles generated from Namespaces and ServiceAccounts
- v3 resources backed by CRDs NOT created via the apiserver or calicoctl (since those two tools generate a truly random UUID which is stored in the annotation)